pasha/functor.py

```diff
@@ -0,0 +1,286 @@
+
+        Returns:
+            (Functor) Functor wrapping the given value or the same value
+                if already a subtype of Functor.
+
+        Raises:
+            ValueError: If no or more than one default functor types
+                could be applied.
+        """
+
+        if isinstance(value, cls):
+            return value
+
+        if getattr(value, '_pasha_functor_'):
+            return value._pasha_functor_()
+
+        functor = None
+
+        for functor_type in cls._functor_types:
+            cur_functor = functor_type.wrap(value)
+
+            if cur_functor is not None:
+                if functor is not None:
+                    raise ValueError('default functor is ambiguous')
+
+                functor = cur_functor
+
+        if functor is None:
+            raise ValueError(f'no default functor for {type(value)}')
+
+        return functor
+
+    @classmethod
+    def wrap(self, value):
+        """Wrap value in this functor type, if possible.
+
+        Args:
+            value (Any): Value to wrap for map operation.
+
+        Returns:
+            (Functor) Functor if wrapping is possible or None.
+        """
+
+        return
+
+    def split(self, num_workers):
+        """Split this functor into work units.
+
+        The values contained in the returned Iterable are passed to this
+        functor's iterate method later on. It may consist of any value
+        suitable to describe each work unit, e.g. an iterable of indices
+        of a sequence.
+
+        Args:
+            num_workers (int): Number of workers processing this
+                functor.
+
+        Returns:
+            (Iterable) Iterable of elements for each work unit.
+        """
+
+        raise NotImplementedError('split')
+
+    def iterate(self, share):
+        """Iterate over a share of this functor.
+
+        Args:
+            share (Any): Element of the Iterable returned by to iterate
+                over.
+
+        Returns:
+            None
+        """
+
+        raise NotImplementedError('iterate')
+
+
+class SequenceFunctor(Functor):
+    """Functor wrapping a sequence.
+
+    This functor can wrap any indexable collection, e.g. list, tuples,
+    or any other type implementing __getitem__. It automatically wraps
+    any value implementing the collections.abc.Sequence type. The kernel
+    is passed the current index and sequence value at that index.
+    """
+
+    def __init__(self, sequence):
+        """Initialize a sequence functor.
+
+        Args:
+            sequence (Sequence): Sequence to process.
+        """
+
+        self.sequence = sequence
+
+    @classmethod
+    def wrap(cls, value):
+        if isinstance(value, Sequence):
+            return cls(value)
+
+    def split(self, num_workers):
+        return np.array_split(np.arange(len(self.sequence)), num_workers)
+
+    def iterate(self, indices):
+        for index in indices:
+            yield index, self.sequence[index]
```

Comment:
Is it worth doing this with slices instead of indices? Especially for the Ndarray case: `arr[slice]` returns a view of the existing memory, whereas `arr[indices]` makes a copy.

Reply:
That's what the `NdarrayFunctor` is for. The `SequenceFunctor` implementation assumes a bare minimum implementation of `__getitem__`, which may not have support for slices. You think one could expect that of any sequence?

Reply:
The NdarrayFunctor inherits the same pattern, though - `.split()` returns indices and `.iterate()` expects indices.

All the sequences I can think of support slicing - lists, tuples, str, bytes, and various kinds of arrays, including h5py datasets. It's such a fundamental part of Python sequences that I doubt any implementation would get far without it. So I think you're probably over-generalising. If we find some case where slicing doesn't work, I'd create a special functor class for that instead.

Reply:
Doh, just learned that using a list of indices actually copies the data and does not return a view. I'll rework this to construct slices instead!

Reply:
Okay, all functors now use slice objects rather than index lists or even loops. Feels quite a bit more elegant than the previous solution, and the `NdarrayFunctor` / `DataArrayFunctor` can use the same code as the `SequenceFunctor` now.

Care to take a look?

Reply:
Yup, I like the new model of dividing sequences (including arrays) using slices. :+1: Sorry I didn't say this explicitly before!